File: apps/api/src/schemas/card.py
"""Card-related Pydantic schemas."""

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class AttackSchema(BaseModel):
    """Schema for card attack data."""

    model_config = ConfigDict(extra="allow")

    name: str
    cost: list[str]
    damage: str | None = None
    effect: str | None = None


class AbilitySchema(BaseModel):
    """Schema for card ability data."""

    model_config = ConfigDict(extra="allow")

    name: str
    type: str | None = None
    effect: str | None = None


class WeaknessResistanceSchema(BaseModel):
    """Schema for weakness/resistance data."""

    model_config = ConfigDict(extra="allow")

    type: str
    value: str


class SetSummaryResponse(BaseModel):
    """Summary schema for card set data."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    series: str
    release_date: date | None = None
    logo_url: str | None = None
    symbol_url: str | None = None


class CardSummaryResponse(BaseModel):
    """Summary schema for card data (used in lists)."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    supertype: str
    types: list[str] | None = None
    set_id: str
    rarity: str | None = None
    image_small: str | None = None


class CardResponse(BaseModel):
    """Full schema for card data."""

    model_config = ConfigDict(from_attributes=True)

    # Identifiers
    id: str
    local_id: str
    name: str
    japanese_name: str | None = None

    # Card type
    supertype: str
    subtypes: list[str] | None = None
    types: list[str] | None = None

    # Pokemon stats
    hp: int | None = None
    stage: str | None = None
    evolves_from: str | None = None
    evolves_to: list[str] | None = None

    # Game mechanics
    attacks: list[dict[str, Any]] | None = None
    abilities: list[dict[str, Any]] | None = None
    weaknesses: list[dict[str, Any]] | None = None
    resistances: list[dict[str, Any]] | None = None
    retreat_cost: int | None = None
    rules: list[str] | None = None

    # Set info
    set_id: str
    rarity: str | None = None
    number: str | None = None

    # Images
    image_small: str | None = None
    image_large: str | None = None

    # Legality
    regulation_mark: str | None = None
    legalities: dict[str, bool] | None = None

    # Timestamps
    created_at: datetime
    updated_at: datetime

    # Nested relationships (optional, for expanded responses)
    set: SetSummaryResponse | None = None

File: packages/shared-types/src/card.ts
/**
 * Card-related TypeScript types matching API schemas (snake_case).
 * These types match the JSON responses from the FastAPI backend.
 */

export interface ApiAttack {
  name: string;
  cost: string[];
  damage?: string | null;
  effect?: string | null;
}

export interface ApiAbility {
  name: string;
  type?: string | null;
  effect?: string | null;
}

export interface ApiWeaknessResistance {
  type: string;
  value: string;
}

export interface ApiSetSummary {
  id: string;
  name: string;
  series: string;
  release_date?: string | null;
  logo_url?: string | null;
  symbol_url?: string | null;
}

export interface ApiCardSummary {
  id: string;
  name: string;
  supertype: string;
  types?: string[] | null;
  set_id: string;
  rarity?: string | null;
  image_small?: string | null;
}

export interface ApiCard {
  // Identifiers
  id: string;
  local_id: string;
  name: string;
  japanese_name?: string | null;

  // Card type
  supertype: string;
  subtypes?: string[] | null;
  types?: string[] | null;

  // Pokemon stats
  hp?: number | null;
  stage?: string | null;
  evolves_from?: string | null;
  evolves_to?: string[] | null;

  // Game mechanics
  attacks?: ApiAttack[] | null;
  abilities?: ApiAbility[] | null;
  weaknesses?: ApiWeaknessResistance[] | null;
  resistances?: ApiWeaknessResistance[] | null;
  retreat_cost?: number | null;
  rules?: string[] | null;

  // Set info
  set_id: string;
  rarity?: string | null;
  number?: string | null;

  // Images
  image_small?: string | null;
  image_large?: string | null;

  // Legality
  regulation_mark?: string | null;
  legalities?: Record<string, boolean> | null;

  // Timestamps
  created_at: string;
  updated_at: string;

  // Nested relationships
  set?: ApiSetSummary | null;
}

export interface ApiCardArchetypeUsage {
  archetype: string;
  inclusion_rate: number;
  avg_copies: number;
}

export interface ApiCardArchetypeUsageResponse {
  card_id: string;
  archetype_usage: ApiCardArchetypeUsage[];
}
